Add pgrag doc (#2397)
neondatabase/website: content/docs/extensions/pgrag.md

---
title: The pgrag extension
subtitle: Create end-to-end Retrieval-Augmented Generation (RAG) pipelines
enableTableOfContents: true
updatedOn: '2024-06-14T07:55:54.371Z'
---

<InfoBlock>
<DocsList title="What you will learn:">
<p>What is RAG?</p>
<p>What's included in a RAG pipeline?</p>
<p>`pgrag` functions</p>
<p>How to use `pgrag`</p>
</DocsList>

<DocsList title="Related docs" theme="docs">
  <a href="https://neon.tech/docs/extensions/pgvector">The pgvector extension</a>
</DocsList>

<DocsList title="Source code" theme="repo">
  <a href="https://github.com/neondatabase-labs/pgrag">pgrag GitHub repository</a>
</DocsList>

</InfoBlock>

> The `pgrag` extension and its accompanying model extensions are designed for creating end-to-end Retrieval-Augmented Generation (RAG) pipelines without leaving your SQL client. No programming languages or libraries are required. With functions provided by `pgrag` and a Postgres database with `pgvector`, you can build a complete RAG pipeline via SQL.

Reviewer: ```suggestion
The `pgrag` extension and its accompanying model extensions are designed for creating end-to-end Retrieval-Augmented Generation (RAG) pipelines without leaving your SQL client. No additional programming languages or libraries are required. With functions provided by `pgrag` and a Postgres database with `pgvector`, you can build a complete RAG pipeline via SQL.
```

<Admonition type="info" title="Experimental Feature">
The `pgrag` extension is experimental and actively being developed. Use it with caution as functionality may change.
</Admonition>

## What is RAG?

> **RAG stands for Retrieval-Augmented Generation**. It's the search for information relevant to a question that includes information alongside the question in a prompt to an AI chat model; for example: "_ChatGPT, please answer questions x using information Y_".

Reviewer: ```suggestion
**RAG stands for Retrieval-Augmented Generation**. It's the search for information relevant to a question that includes information alongside the question in a prompt to an AI chat model. For example, "_ChatGPT, please answer questions x using information Y_".
```

---

> ## What's included in a RAG pipeline?

Reviewer: In a later update, might be good to show this visually in a flow diagram.

There are two main stages in a RAG pipeline:

1. **Preparing and indexing the information**: This stage involves:
   - Loading documents and extracting text
   - Splitting the documents into chunks
   - Generating embeddings for the chunks
   - Storing the embeddings alongside chunks in your vector database
2. **Handling incoming questions**: This second stage involves:
   - Vectorizing the question to create an embedding
   - Using the question embedding to find relevant document chunks based on the shortest vector distances
   - Retrieving document chunks from the database
   - Reranking the chunks for the best-matching ones
   - Prompting the generative AI chat model with the question and relevant document chunks
   - Generating the answer

---

## What does pgrag support?

With the exception of storing and retrieving embeddings, which is supported by Postgres with `pgvector`, `pgrag` supports all of the steps listed above. Specifically, `pgrag` supports:

- **Text extraction and conversion**

>   - Simple text extraction from PDF documents (using [pdf-extract](https://github.com/jrmuizel/pdf-extract)). Currently, there is no OCR or support for complex layout and formatting.

Reviewer: ```suggestion
  - Simple text extraction from PDF documents (using [pdf-extract](https://github.com/jrmuizel/pdf-extract)). Currently, there is no Optical Character Recognition (OCR) or support for complex layout and formatting.
```

- **Text extraction**

  These functions extract text from PDFs, Word files, and HTML.

  - `rag.text_from_pdf(bytea) -> text`
  - `rag.text_from_docx(bytea) -> text`
  - `rag.markdown_from_html(text) -> text`

- **Splitting text into chunks**

  These functions split the extracted text into chunks by character count or token count.

  - `rag.chunks_by_character_count(text, max_chars, overlap) -> text[]`
  - `rag_bge_small_en_v15.chunks_by_token_count(text, max_tokens, overlap) -> text[]`

- **Generating embeddings for chunks**

  These functions generate embeddings for chunks either directly in the extension using a small but best-in-class model on the database server or by calling out to a 3rd-party API such as OpenAI.

  - `rag_bge_small_en_v15.embedding_for_passage(text) -> vector(384)`
  - `rag.openai_text_embedding_3_small(text) -> vector(1536)`

- **Generating embeddings for questions**

  These functions generate embeddings for the questions.

  - `rag_bge_small_en_v15.embedding_for_query(text) -> vector(384)`
  - `rag.openai_text_embedding_3_small(text) -> vector(1536)`

- **Reranking**

>   This function reranks chunks against the question using a small but best-in-class model that runs locally on the database server.

Reviewer: ```suggestion
  This function reranks chunks against the question using a small but best-in-class model that runs locally on your Postgres server.
```

  - `rag_jina_reranker_v1_tiny_en.rerank_distance(text, text) -> real`

- **Calling out to chat models**

  This function makes API calls to AI chat models such as ChatGPT to generate an answer using the question and the chunks together.

  - `rag.openai_chat_completion(json) -> json`

---

## End-to-end RAG example

**1. Create a `docs` table and ingest some PDF documents as text**

```sql
drop table docs cascade;
create table docs
( id int primary key generated always as identity
, name text not null
, fulltext text not null
);

\set contents `base64 < /path/to/first.pdf`
insert into docs (name, fulltext)
values ('first.pdf', rag.text_from_pdf(decode(:'contents','base64')));

\set contents `base64 < /path/to/second.pdf`
insert into docs (name, fulltext)
values ('second.pdf', rag.text_from_pdf(decode(:'contents','base64')));

\set contents `base64 < /path/to/third.pdf`
insert into docs (name, fulltext)
values ('third.pdf', rag.text_from_pdf(decode(:'contents','base64'))));
```

> **2. Ceate an `embeddings` table, chunk the text, and generate embeddings for the chunks (performed locally)**

Reviewer: ```suggestion
**2. Create an `embeddings` table, chunk the text, and generate embeddings for the chunks (performed locally)**
```
